mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendWelcomeEmailAction.php

```diff
@@ -0,0 +1,119 @@
+    $subscriberSubject = $workflowRun->getSubjects()['mailpoet:subscriber'] ?? null;
+    if (!$subscriberSubject instanceof SubscriberSubject) {
+      throw InvalidStateException::create()->withMessage(__('No mailpoet:subscriber subject provided.', 'mailpoet'));
+    }
+
+    $segmentSubject = $workflowRun->getSubjects()['mailpoet:segment'] ?? null;
+    if (!$segmentSubject instanceof SegmentSubject) {
+      throw InvalidStateException::create()->withMessage(__('No mailpoet:segment subject provided.', 'mailpoet'));
+    }
+
+    $globalSubscriberStatus = $subscriberSubject->getSubscriberStatusField()->getFactory()();
+    if ($globalSubscriberStatus !== SubscriberEntity::STATUS_SUBSCRIBED) {
+      throw InvalidStateException::create()->withMessage(__(sprintf("Cannot send a welcome email to a subscriber with a global subscription status of '%s'.", $globalSubscriberStatus), 'mailpoet'));
+    }
+
+    $segmentName = $segmentSubject->getNameField()->getFactory()();
+    $idField = $subscriberSubject->getSubscriberIdField();
+    $subscriberId = $idField->getFactory()();
+    if ($subscriberId === null) {
+      throw NotFoundException::create()->withMessage(__(sprintf("Subscriber with ID '%s' not found.", $subscriberId), 'mailpoet'));
+    }
+    
+    $subscriber = $this->subscribersRepository->findOneById($subscriberId);
+    // This is for PHPStan, which doesn't understand that retrieving the ID above means the subscriber entity exists
+    if (!$subscriber instanceof SubscriberEntity) {
+      throw NotFoundException::create()->withMessage(__(sprintf("Subscriber with ID '%s' not found.", $subscriberId), 'mailpoet'));
+    }
+
+    // TODO: Are these necessarily "subscribed" status? Maybe there's a better way to check this
+    $subscriberSegments = $subscriber->getSegments();
+    $subscribedSegmentNames = array_map(function(SegmentEntity $segment) {
+      return $segment->getName();
+    }, $subscriberSegments);
+    
+    if (!in_array($segmentName, $subscribedSegmentNames)) {
+      throw InvalidStateException::create()->withMessage(__(sprintf("Subscriber with ID '%s' is not subscribed to segment '%s'.", $subscriberId, $segmentName), 'mailpoet'));
+    }
+    
+    $welcomeEmailId = (int)$step->getArgs()['welcomeEmailId'];
+    $newsletter = $this->newslettersRepository->findOneById($welcomeEmailId);
+    if ($newsletter === null) {
+      throw NotFoundException::create()->withMessage(__(sprintf("Welcome Email with ID '%s' not found.", $welcomeEmailId), 'mailpoet'));
+    }
+
+    // This check also occurs in createWelcomeNotificationSendingTask, in which case the method returns null, but
+    // that's not the only thing that causes a return value of null, so let's check here so we can craft a more
+    // meaningful exception.
+    $previouslyScheduledNotification = $this->scheduledTasksRepository->findByNewsletterAndSubscriberId($newsletter, $subscriberId);
```

Comment:
I couldn't see a way around needing a subscriber's ID in this action in large part because of this check and the actual creation of the sending task on L106, which is why I added the ID field and avoided using the email field.

Reply:
Makes sense, since it's a "welcome" email, subscriber should always exist.